thorlcr/activities/pipewrite/thpwslave.cpp

```diff
@@ -63,75 +63,88 @@ class CPipeWriteSlaveActivity : public ProcessSlaveActivity
                 {
-                    char error[512];
-                    size32_t sz = pipe->readError(sizeof(error), error);
-                    if (sz && sz!=(size32_t)-1) 
-                        stdError.append(", stderr: '").append(sz, error).append("'");
-                }
-                catch (IException *e)
-                {
-                    ActPrintLog(e, "Error reading pipe stderr");
-                    e->Release();
+                    try
+                    {
+                        char error[512];
+                        size32_t sz = pipe->readError(sizeof(error), error);
+                        if (sz && sz!=(size32_t)-1)
+                            stdError.append(", stderr: '").append(sz, error).append("'");
+                    }
+                    catch (IException *e)
+                    {
+                        ActPrintLog(e, "Error reading pipe stderr");
+                        e->Release();
+                    }
                 }
+                if (START_FAILURE == retcode) // PIPE process didn't start at all, START_FAILURE is our own error code
+                    throw MakeActivityException(this, TE_PipeReturnedFailure, "Process failed to start: %s - PIPE(%s)", stdError.str(), pipeCommand.get());
+                else
+                    throw MakeActivityException(this, TE_PipeReturnedFailure, "Process returned %d:%s - PIPE(%s)", retcode, stdError.str(), pipeCommand.get());
             }
-            throw MakeActivityException(this, TE_PipeReturnedFailure, "Process returned %d:%s - PIPE(%s)", retcode, stdError.str(), pipeCommand.get());    
+            pipe.clear();
         }
     }
     void process()
     {
         start();
-        if (!writeTransformer)
-        {
-            writeTransformer.setown(createPipeWriteXformHelper(helper->getPipeFlags(), helper->queryXmlOutput(), helper->queryCsvOutput(), ::queryRowInterfaces(input)->queryRowSerializer()));
-            writeTransformer->ready();
-        }
-        processed = THORDATALINK_STARTED;
+        Owned<IPipeProcessException> pipeException;
+
+        ActPrintLog("process");
         try
         {
-            ActPrintLog("process");
-
-            open(); 
+            if (!writeTransformer)
+            {
+                writeTransformer.setown(createPipeWriteXformHelper(helper->getPipeFlags(), helper->queryXmlOutput(), helper->queryCsvOutput(), ::queryRowInterfaces(input)->queryRowSerializer()));
+                writeTransformer->ready();
+            }
+            processed = THORDATALINK_STARTED;
+            open();
             write();
+            if (!recreate)
+                closePipe();
         }
-        catch (IException *e)
+        catch (IPipeProcessException *e)
         {
-            try { close(); } catch (CATCHALL) { ActPrintLog("Error closing file"); }
-            ActPrintLog(e, "exception");
-            throw;
+            pipeException.setown(e);
         }
```

Comment:
Should we add another catch for any other possible exceptions ?

Reply:
I don't think needed. For any other exception, we want it to just propagate normally.
If it's a pipe exception, we want to ensure verifyPipe() is called, so that a stderr exception can be generated.